hwy/ops/ppc_vsx-inl.h

```diff
@@ -3731,44 +3923,85 @@ HWY_API Vec64<uint32_t> DemoteTo(D /* tag */, Vec128<double> v) {
   const Rebind<uint64_t, D> du64;
   const Vec128<uint64_t> vu64 = BitCast(du64, f64_to_u32);
   return Vec64<uint32_t>{vec_pack(vu64.raw, vu64.raw)};
+#endif
+}
+
+#if HWY_S390X_HAVE_Z14
+namespace detail {
+
+template <class V, HWY_IF_I64(TFromV<V>)>
+HWY_INLINE VFromD<RebindToFloat<DFromV<V>>> ConvToF64WithRoundToOdd(V v) {
+  __vector double raw_result;
+  // Use inline assembly to do a round-to-odd I64->F64 conversion on Z14
+  __asm__("vcdgb %0,%1,0,3" : "=v"(raw_result) : "v"(v.raw));
+  return VFromD<RebindToFloat<DFromV<V>>>{raw_result};
+}
+
+template <class V, HWY_IF_U64(TFromV<V>)>
+HWY_INLINE VFromD<RebindToFloat<DFromV<V>>> ConvToF64WithRoundToOdd(V v) {
+  __vector double raw_result;
+  // Use inline assembly to do a round-to-odd U64->F64 conversion on Z14
+  __asm__("vcdlgb %0,%1,0,3" : "=v"(raw_result) : "v"(v.raw));
+  return VFromD<RebindToFloat<DFromV<V>>>{raw_result};
 }
 
+}  // namespace detail
+#endif  // HWY_S390X_HAVE_Z14
+
 template <class D, HWY_IF_V_SIZE_D(D, 4), HWY_IF_F32_D(D)>
-HWY_API Vec32<float> DemoteTo(D /* tag */, Vec64<int64_t> v) {
+HWY_API Vec32<float> DemoteTo(D df32, Vec64<int64_t> v) {
+#if HWY_S390X_HAVE_Z14
+  return DemoteTo(df32, detail::ConvToF64WithRoundToOdd(v));
```

Comment:
I'm curious why round to odd?

Reply:
Doing an I64->F64 or U64->F32 conversion using round to odd ensures that the I64->F32 or U64->F32 DemoteTo conversion produces the same result as `static_cast<float>(v[i])`.

A round-to-odd integer to floating-point conversion is equivalent to a truncating integer to floating-point conversion followed by a bitwise OR of the bit representation by 1 if the truncated floating-point value is not equal to the original value.

The SSE2/SSSE3/SSE4/AVX2/A64 NEON/WASM UI64->F32 DemoteTo implementations first do a round-to-odd conversion of `v[i]` to a F64 followed by a demotion of the rounded-to-odd F64 conversion result to float to ensure that SSE2/SSSE3/SSE4/AVX2/A64 NEON/WASM UI64->F32 matches the behavior of AVX3/PPC/RVV/SVE UI64->F32 DemoteTo.

A round-to-odd I64/U64->F64 conversion avoids the double-rounding that can occur with `static_cast<float>(static_cast<double>(v[i]))` or `DemoteTo(df32, ConvertTo(df64, v))`.

Here is the round-to-odd I64->F64 conversion snippet from the SSE2/SSSE3/SSE4/AVX2 I64->F32 DemoteTo implementation in hwy/ops/x86_128-inl.h:
```
  const auto k2p64_63 = Set(df64, 27670116110564327424.0);
  const auto f64_hi52 =
      Xor(BitCast(df64, ShiftRight<12>(BitCast(du64, v))), k2p64_63) - k2p64_63;
  const auto f64_lo12 =
      PromoteTo(df64, BitCast(di32, And(TruncateTo(du32, BitCast(du64, v)),
                                        Set(du32, uint32_t{0x00000FFF}))));

  const auto f64_sum = f64_hi52 + f64_lo12;
  const auto f64_carry = (f64_hi52 - f64_sum) + f64_lo12;

  const auto f64_sum_is_inexact =
      ShiftRight<63>(BitCast(du64, VecFromMask(df64, f64_carry != Zero(df64))));
  const auto f64_bits_decrement =
      And(ShiftRight<63>(BitCast(du64, Xor(f64_sum, f64_carry))),
          f64_sum_is_inexact);

  const auto adj_f64_val = BitCast(
      df64,
      Or(BitCast(du64, f64_sum) - f64_bits_decrement, f64_sum_is_inexact));
```

Z14 can carry out the UI64->F64 round-to-odd conversion using a single vcdgb or vcdlgb instruction, whereas the SSE2 UI64->F64 round-to-odd conversion using the above algorithm requires 18 instructions.

Reply:
I see, thank you. That's handy :)